Improve Player Lifecycle Docs (#869)
spatialos/gdk-for-unity: docs/modules/player-lifecycle/custom-player-creation.md

<%(TOC)%>
# Customising player creation

<%(Callout message="
Before reading this document, make sure you are familiar with:

  * [Basic player spawning]({{urlRoot}}/modules/player-lifecycle/basic-player-creation)
  * [Creating entity templates]({{urlRoot}}/reference/concepts/entity-templates)
  * [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
")%>

## Manually sending player creation requests

> By default, the Player Lifecycle Feature Module automatically sends a player creation request is sent as soon as the client-worker instance connects to SpatialOS.

Reviewer: ~automatically~

To modify this behaviour, you must set `PlayerLifecycleConfig.AutoRequestPlayerCreation` to false by one of the following ways:

* Add an extra `false` argument when adding client systems: `PlayerLifecycleHelper.AddClientSystems(Worker.World, autoRequestPlayerCreation: false);`
* Set the field to false manually during set-up: `PlayerLifecycleConfig.AutoRequestPlayerCreation = false;`

> * Directly modify `PlayerLifecycleConfig` to be false by default

Reviewer: Does this involve a "fork"? If so we shouldn't mention it

Once `AutoRequestPlayerCreation` is set to false you must manually call the `RequestPlayerCreation` method, which can be found in the `SendCreatePlayerRequestSystem`.

```csharp
var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
playerCreationSystem.RequestPlayerCreation();
```

To find out more about the `RequestPlayerCreation` method, take a look at the `SendCreatePlayerRequestSystem` [API reference documentation]({{urlRoot}}/api/player-lifecycle/send-create-player-request-system#methods).

## Adding a player creation callback

If manual player creation is enabled, you can configure a callback to be triggered when a player creation response is received. The callback must be of type `Action<PlayerCreator.CreatePlayer.ReceivedResponse>`.

For example:

```csharp
void CallPlayerCreation()
{
    var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
    playerCreationSystem.RequestPlayerCreation(callback: OnCreatePlayerResponse);
}

> void OnCreatePlayerResponse(PlayerCreator.CreatePlayer.ReceivedResponse response)

Reviewer: Lets add access modifiers on these :)

{
    if (response.StatusCode != StatusCode.Success)
    {
        Debug.LogWarning($"Error: {response.Message}");
    }
}
```

## Using arbitrary data in the player creation loop

> The default player creation loop sends requests automatically. To use arbitrary data in the loop, you must ensure that `AutoRequestPlayerCreation` is set to false. Some

Reviewer: Some

> 1. Disable `AutoRequestPlayerCreation`, some ways of doing this are described [here](#manually-sending-player-creation-requests).

Reviewer: You can just say:
```
Disable `AutoRequestPlayerCreation` as described in the [step above](#link) 
```

1. Call `RequestPlayerCreation` with serialized data.
2. Deserialize data back into the original type.

##### Call RequestPlayerCreation with serialized data

> Then, call `RequestPlayerCreation`. Ensure that your arbitrary data is serialized into a byte array.

Reviewer: There's no previous step strictly, `Then` reads oddly.

```csharp
var myArguments = new SampleArgumentsObject
{
  PlayerName = "playerName",
  SpawnPosition = new Coordinates(50, 0, 75))
};
var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
var serializedArguments = SerializeArguments(myArguments);
playerCreationSystem.RequestPlayerCreation(serializedArguments);
```

> Where an example implementation of `SerializeArguments` could be:

Reviewer: This could be in an expandable with a header of:

"How do I serialize my data?" 

And we can say - there are many ways to serialize data, an example is blah blah blah

```csharp
private byte[] SerializeArguments(object playerCreationArguments)
{
    using (var memoryStream = new MemoryStream())
    {
        var binaryFormatter = new BinaryFormatter();
        binaryFormatter.Serialize(memoryStream, playerCreationArguments);
        return memoryStream.ToArray();
    }
}
```

##### Deserialize data into the same type you originally serialized from

Lastly, ensure that you are deserializing the byte array into the same type of object you serialized it from. For example:

> ```csharp

Reviewer: There's a lot of code in here that isn't really relevant for a user.

The main bit of information that they care about is that the `playerCreationArguments` is the serialized byte array in the previous step and is given when your EntityTemplate delegate is called. That is never explicitly stated in the docs.

Author: Moved the serialize/deserialize sample code into expandables

I think that is good to continue the example from earlier with the `SampleArgumentsObject` so users have some example of how they could make use of the `byte[]` field

<%(TOC)%>
# Customising player creation

<%(Callout message="
Before reading this document, make sure you are familiar with:

  * [Basic player spawning]({{urlRoot}}/modules/player-lifecycle/basic-player-creation)
  * [Creating entity templates]({{urlRoot}}/reference/concepts/entity-templates)
  * [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
")%>

## Manually sending player creation requests

By default, the Player Lifecycle Feature Module sends a player creation request is sent as soon as the client-worker instance connects to SpatialOS.

To modify this behaviour, you must set `PlayerLifecycleConfig.AutoRequestPlayerCreation` to false by one of the following ways:

* Add an extra `false` argument when adding client systems: `PlayerLifecycleHelper.AddClientSystems(Worker.World, autoRequestPlayerCreation: false);`
* Set the field to false manually during set-up: `PlayerLifecycleConfig.AutoRequestPlayerCreation = false;`

> Once `AutoRequestPlayerCreation` is set to false you must manually call the `RequestPlayerCreation` method, which can be found in the `SendCreatePlayerRequestSystem`.

Reviewer: Once `AutoRequestPlayerCreation` is set to false, you can call the `SendCreatePlayerRequestSystem.RequestPlayerCreation()` method when you are ready to request player spawning.

```csharp
var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
playerCreationSystem.RequestPlayerCreation();
```

To find out more about the `RequestPlayerCreation` method, take a look at the `SendCreatePlayerRequestSystem` [API reference documentation]({{urlRoot}}/api/player-lifecycle/send-create-player-request-system#methods).

> ## Adding a player creation callback

Reviewer: Should this be a sub-header of the above?

If manual player creation is enabled, you can configure a callback to be triggered when a player creation response is received. The callback must be of type `Action<PlayerCreator.CreatePlayer.ReceivedResponse>`.

For example:

```csharp
public void CallPlayerCreation()
{
    var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
    playerCreationSystem.RequestPlayerCreation(callback: OnCreatePlayerResponse);
}

private void OnCreatePlayerResponse(PlayerCreator.CreatePlayer.ReceivedResponse response)
{
    if (response.StatusCode != StatusCode.Success)
    {
        Debug.LogWarning($"Error: {response.Message}");
    }
}
```

> ## Using arbitrary data in the player creation loop

Reviewer: Similarly, should this be a sub header? Not entirely sure

Author: I don't think it should because there is more to this than calling `RequestPlayerCreation`. Even if there is some overlap, it probably makes sense to keep the information about the `playerCreationArguments`/how to serialize/deserialize under a separate header

To send arbitrary data during the creation of the player, you need to manually send the player creation request [as described earlier](#manually-sending-player-creation-requests).

##### Call RequestPlayerCreation with serialized data

> Get the `SendCreatePlayerRequestSystem`, and invoke the `RequestPlayerCreation` method. Ensure that your arbitrary data is serialized into a byte array.

Reviewer: Lets try to keep the phrasing similar wrt `RequestPlayerCreation` with the parts above.

<%(TOC)%>
# Customising player creation

> <%(Callout message="

Reviewer 2: why is this in a callout?

Author: I like how this looks on the page + it grabs more attention than normal text

Reviewer: I think the callout is perfect for this and is miles better than what we currently do: 

![image](https://user-images.githubusercontent.com/13353733/55500913-3dba6700-5641-11e9-9b3f-e78ebb93ffd9.png)

Author: compared to
![image](https://user-images.githubusercontent.com/10051819/55501122-9f7ad100-5641-11e9-886e-e0a396118266.png)

Reviewer 2: hm.... as long as we make sure that there are not too many callouts, cause then the page is just really colorful

Reviewer 2: and we should be consistent throughout all pages :P

Before reading this document, make sure you are familiar with:

  * [Basic player spawning]({{urlRoot}}/modules/player-lifecycle/basic-player-creation)
  * [Creating entity templates]({{urlRoot}}/reference/concepts/entity-templates)
  * [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
")%>

> ## Manually sending player creation requests

Reviewer 2: would prefer to rephrase these titles to be "How to ..." instead of using the "-ing" form of verbs. I believe we removed that from most docs at some point. so for example

```suggestion
## How to manually send player creation requests
```

> By default, the Player Lifecycle Feature Module sends a player creation request is sent as soon as the client-worker instance connects to SpatialOS.

Reviewer 2: ```suggestion
By default, the Player Lifecycle Feature Module sends a player creation request soon as the client-worker instance connects to SpatialOS.
```

To modify this behaviour, you must set `PlayerLifecycleConfig.AutoRequestPlayerCreation` to false by one of the following ways:

* Add an extra `false` argument when adding client systems: `PlayerLifecycleHelper.AddClientSystems(Worker.World, autoRequestPlayerCreation: false);`

> * Set the field to false manually during set-up: `PlayerLifecycleConfig.AutoRequestPlayerCreation = false;`

Reviewer 2: ```suggestion
* Set the field to false during set-up: `PlayerLifecycleConfig.AutoRequestPlayerCreation = false;`
```

> Once `AutoRequestPlayerCreation` is set to false you must manually call the `RequestPlayerCreation` method, which can be found in the `SendCreatePlayerRequestSystem`.

Reviewer 2: ```suggestion
To spawn the player after disabling the default behaviour, you need to call the `RequestPlayerCreation` method. This method can be found in the `SendCreatePlayerRequestSystem`.
```

```csharp
var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
playerCreationSystem.RequestPlayerCreation();
```

To find out more about the `RequestPlayerCreation` method, take a look at the `SendCreatePlayerRequestSystem` [API reference documentation]({{urlRoot}}/api/player-lifecycle/send-create-player-request-system#methods).

## Adding a player creation callback

If manual player creation is enabled, you can configure a callback to be triggered when a player creation response is received. The callback must be of type `Action<PlayerCreator.CreatePlayer.ReceivedResponse>`.

For example:

```csharp
public void CallPlayerCreation()
{
    var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
    playerCreationSystem.RequestPlayerCreation(callback: OnCreatePlayerResponse);
}

private void OnCreatePlayerResponse(PlayerCreator.CreatePlayer.ReceivedResponse response)
{
    if (response.StatusCode != StatusCode.Success)
    {
        Debug.LogWarning($"Error: {response.Message}");
    }
}
```

## Using arbitrary data in the player creation loop

> The default player creation loop sends requests automatically. To use arbitrary data in the loop, you must ensure that `AutoRequestPlayerCreation` is set to false.

Reviewer 2: ```suggestion
To send arbitrary data during the creation of the player, you need to [manually send the player creation request](#manually-sending-player-creation-requests):
```

and then add a code example of how to do this

1. Disable `AutoRequestPlayerCreation` as described in the [step above](#manually-sending-player-creation-requests).
1. Call `RequestPlayerCreation` with serialized data.

> 2. Deserialize data back into the original type.

Reviewer 2: remove this list and use code examples

Reviewer 2: ah just saw that you already added code examples, in which case. just remove the list

##### Call RequestPlayerCreation with serialized data

Get the `SendCreatePlayerRequestSystem`, and invoke the `RequestPlayerCreation` method. Ensure that your arbitrary data is serialized into a byte array.

```csharp
var myArguments = new SampleArgumentsObject
{
  PlayerName = "playerName",
  SpawnPosition = new Coordinates(50, 0, 75))
};
var playerCreationSystem = World.GetExistingManager<SendCreatePlayerRequestSystem>();
var serializedArguments = SerializeArguments(myArguments);
playerCreationSystem.RequestPlayerCreation(serializedArguments);
```

<%(#Expandable title="How do I serialize my data?")%>

There are numerous ways to serialize data. An example implementation could be:

```csharp
private byte[] SerializeArguments(object playerCreationArguments)
{
    using (var memoryStream = new MemoryStream())
    {
        var binaryFormatter = new BinaryFormatter();
        binaryFormatter.Serialize(memoryStream, playerCreationArguments);
        return memoryStream.ToArray();
    }
}
```

<%(/Expandable)%>

##### Deserialize data into the same type you originally serialized from

> The data serialized in the previous step on the client-worker is then sent across to the server-worker as part of the player creation request. It is then passed in as the `playerCreationArguments` parameter of the entity template delegate that is invoked.

Reviewer 2: ```suggestion
The data serialized in the previous step is sent across to the server-worker as part of the player creation request. This data is passed in as the `playerCreationArguments` parameter of the entity template delegate that is invoked.
```
